File: typescript/src/clinical.ts
/**
 * Clinical trial blinding/unblinding types — wire shapes for `v1-clinical-unblind`.
 *
 * Supports ICH E6(R2) §4.8 / 21 CFR Part 11 §11.10(a) / §11.50 / §11.300
 * execution-time authorization of clinical trial unblinding operations.
 * The blinding state machine is append-only; each transition produces an
 * immutable {@link ClinicalUnblindingEvent} in the audit ledger.
 */

// Status values matching the DB CHECK constraint on `clinical_trial_blinds.status`.
export type ClinicalBlindingStatus =
  | 'blinded'
  | 'unblinding_in_progress'
  | 'unblinded'
  | 'emergency_unblinded'
  | 'suspended';

/**
 * A clinical trial blinding record.
 * Returned by GET list and GET ?trial_id=X.
 */
export interface ClinicalTrialBlind {
  id: string;
  org_id: string;
  trial_id: string;
  trial_name: string;
  protocol_number: string | null;
  phase: string | null;
  blinding_type: string;
  status: ClinicalBlindingStatus;
  established_by: string;
  established_evaluation_id: string | null;
  sponsor_org: string | null;
  randomization_code_hash: string;
  created_at: string;
  unblinded_by: string | null;
  unblinded_at: string | null;
  unblinding_evaluation_id: string | null;
  emergency_unblinded_by: string | null;
  emergency_unblinded_at: string | null;
  suspended_at: string | null;
  suspended_by: string | null;
}

/**
 * A single audit event from the clinical unblinding ledger.
 * Returned by GET /history. Append-only — corrections are new events.
 */
export interface ClinicalUnblindingEvent {
  id: string;
  org_id: string;
  trial_id: string;
  blind_id: string;
  event_type:
    | 'blind_established'
    | 'unblinding_initiated'
    | 'unblinding_executed'
    | 'emergency_unblinding_executed'
    | 'blinding_reinstated'
    | 'suspended'
    | 'reinstated';
  actor_id: string;
  evaluation_id: string | null;
  permit_token_hash: string | null;
  reason: string;
  /** §11.50(a)(2) electronic signature meaning text. */
  approval_meaning: string | null;
  subject_ids: string[] | null;
  emergency: boolean;
  unblinding_scope: string;
  occurred_at: string;
  metadata: Record<string, unknown> | null;
}

/**
 * POST /blind request — establish a new blinding record for a clinical trial.
 */
export interface ClinicalBlindRequest {
  trial_id: string;
  trial_name: string;
  phase: string;
  blinding_type: string;
  /** SHA-256 hex hash of the randomization code. Never the code itself. */
  randomization_code_hash: string;
  established_by: string;
  reason: string;
  protocol_number?: string;
  evaluation_id?: string;
  /** §11.50(a)(2) electronic signature meaning. */
  approval_meaning?: string;
  sponsor_org?: string;
}

/** POST /blind response. */
export interface ClinicalBlindResponse {
  blind: ClinicalTrialBlind;
}

/**
 * POST /unblind request.
 * Server requires `approval_meaning` ≥ 20 chars (§11.50(a)(2)).
 */
export interface ClinicalUnblindRequest {
  trial_id: string;
  actor_id: string;
  reason: string;
  /** §11.50(a)(2) electronic signature meaning; server requires ≥ 20 characters. */
  approval_meaning: string;
  evaluation_id?: string;
  permit_token_hash?: string;
  /** Scope of unblinding (e.g. `"full"`, `"partial"`, `"individual"`). */
  unblinding_scope?: string;
  /** DSMB authorization reference for audit trail. */
  dsmb_authorization_ref?: string;
}

/**
 * POST /emergency request — ICH E6(R2) §4.8 individual-patient emergency unblinding.
 */
export interface ClinicalEmergencyRequest {
  trial_id: string;
  actor_id: string;
  subject_id: string;
  emergency_justification: string;
  evaluation_id?: string;
  permit_token_hash?: string;
  /** §11.50(a)(2) electronic signature meaning. */
  approval_meaning?: string;
}

/**
 * Response body for POST /unblind and POST /emergency.
 */
export interface ClinicalMutationResponse {
  success: boolean;
  trial_id: string;
  status?: ClinicalBlindingStatus;
  /** Present on emergency unblinding — the affected subject identifier. */
  subject_id?: string;
  event_type?: string;
  unblinded_by?: string;
  unblinded_at?: string;
  occurred_at?: string;
  evaluation_id?: string | null;
}

/** GET list response. */
export interface ClinicalTrialListResponse {
  trials: ClinicalTrialBlind[];
}

/** GET ?trial_id=X response. */
export interface ClinicalTrialGetResponse {
  trial: ClinicalTrialBlind;
}

/** GET /history response. */
export interface ClinicalHistoryResponse {
  events: ClinicalUnblindingEvent[];
}

/** POST /verify-permit request. */
export interface ClinicalVerifyPermitRequest {
  trial_id: string;
  permit_token: string;
  /** One of the clinical action classes. */
  action_type:
    | "trial.blinding.setup"
    | "trial.unblinding.execute"
    | "trial.unblinding.emergency";
  actor_id: string;
}

/** POST /verify-permit result (as returned by the permit verifier). */
export interface ClinicalVerifyPermitResult {
  valid?: boolean;
  outcome?: string;
  verify_error_code?: string;
  reason?: string;
}

/** Optional filters for {@link ClinicalTrialsSubClient.list}. */
export interface ClinicalTrialListQuery {
  status?: ClinicalBlindingStatus;
  limit?: number;
  offset?: number;
}

/**
 * Client for the clinical trial blinding/unblinding gate
 * (`v1-clinical-unblind`, public path `/v1/clinical-unblind`). Reads require
 * the `clinical:read` scope; writes require `clinical:manage`.
 */
export interface ClinicalTrialsSubClient {
  list(query?: ClinicalTrialListQuery): Promise<ClinicalTrialListResponse>;
  get(trialId: string): Promise<ClinicalTrialGetResponse>;
  history(trialId: string): Promise<ClinicalHistoryResponse>;
  blind(request: ClinicalBlindRequest): Promise<ClinicalBlindResponse>;
  requestUnblind(request: ClinicalUnblindRequest): Promise<ClinicalMutationResponse>;
  emergencyUnblind(request: ClinicalEmergencyRequest): Promise<ClinicalMutationResponse>;
  verifyPermit(request: ClinicalVerifyPermitRequest): Promise<ClinicalVerifyPermitResult>;
}

const CLINICAL_BASE = "/v1/clinical-unblind";

/**
 * Build the clinical-trials sub-client from the host client's request
 * primitives. Mirrors the other `make*Client` factories.
 */
export function makeClinicalTrialsClient(
  getFn: <T>(path: string, query?: URLSearchParams) => Promise<{ body: T }>,
  postFn: <T>(path: string, body: unknown) => Promise<{ body: T }>,
): ClinicalTrialsSubClient {
  return {
    async list(query = {}) {
      const sp = new URLSearchParams();
      if (query.status) sp.set("status", query.status);
      if (query.limit != null) sp.set("limit", String(query.limit));
      if (query.offset != null) sp.set("offset", String(query.offset));
      const qs = [...sp].length ? sp : undefined;
      const { body } = await getFn<ClinicalTrialListResponse>(CLINICAL_BASE, qs);
      return body;
    },

    async get(trialId: string) {
      const sp = new URLSearchParams({ trial_id: trialId });
      const { body } = await getFn<ClinicalTrialGetResponse>(CLINICAL_BASE, sp);
      return body;
    },

    async history(trialId: string) {
      const sp = new URLSearchParams({ trial_id: trialId });
      const { body } = await getFn<ClinicalHistoryResponse>(`${CLINICAL_BASE}/history`, sp);
      return body;
    },

    async blind(request: ClinicalBlindRequest) {
      const { body } = await postFn<ClinicalBlindResponse>(`${CLINICAL_BASE}/blind`, request);
      return body;
    },

    async requestUnblind(request: ClinicalUnblindRequest) {
      const { body } = await postFn<ClinicalMutationResponse>(`${CLINICAL_BASE}/unblind`, request);
      return body;
    },

    async emergencyUnblind(request: ClinicalEmergencyRequest) {
      const { body } = await postFn<ClinicalMutationResponse>(`${CLINICAL_BASE}/emergency`, request);
      return body;
    },

    async verifyPermit(request: ClinicalVerifyPermitRequest) {
      const { body } = await postFn<ClinicalVerifyPermitResult>(
        `${CLINICAL_BASE}/verify-permit`,
        request,
      );
      return body;
    },
  };
}

File: python/atlasent/clinical.py
"""Clinical trial blinding/unblinding wire types.

Parity with ``typescript/src/clinical.ts``.

Wire surface: ``v1-clinical-unblind`` edge function (atlasent-api).
Operations: ``list`` (GET), ``get`` (GET ?trial_id=X), ``history``
(GET /history), ``blind`` (POST /blind), ``unblind`` (POST /unblind),
``emergency`` (POST /emergency).

Supports ICH E6(R2) §4.8 / 21 CFR Part 11 §11.10(a) / §11.50 / §11.300
execution-time authorization of clinical trial unblinding operations.
The blinding state machine is append-only; each transition produces an
immutable :class:`ClinicalUnblindingEvent` in the audit ledger.

Wire-stable as ``clinical.v1``.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Literals
# ---------------------------------------------------------------------------

ClinicalBlindingStatus = Literal[
    "blinded",
    "unblinding_in_progress",
    "unblinded",
    "emergency_unblinded",
    "suspended",
]
"""Status values matching the DB CHECK on ``clinical_trial_blinds.status``."""

ClinicalUnblindingEventType = Literal[
    "blind_established",
    "unblinding_initiated",
    "unblinding_executed",
    "emergency_unblinding_executed",
    "blinding_reinstated",
    "suspended",
    "reinstated",
]
"""Event types recorded in the append-only ``clinical_unblinding_events`` ledger."""

# Terminal states in which the randomized assignment is (fully or partially) revealed.
_UNBLINDED_STATES: frozenset[str] = frozenset({"unblinded", "emergency_unblinded"})


# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------


class ClinicalTrialBlind(BaseModel):
    """A clinical trial blinding record.

    Returned by ``GET`` (list) and ``GET ?trial_id=X``. The lifecycle is a
    state machine — ``blinded → unblinding_in_progress | emergency_unblinded
    | suspended → unblinded`` — enforced server-side by a BEFORE UPDATE
    trigger; illegal transitions are rejected.
    """

    id: str
    org_id: str
    trial_id: str
    trial_name: str
    protocol_number: str | None = None
    phase: str | None = None
    blinding_type: str
    status: ClinicalBlindingStatus
    established_by: str
    established_evaluation_id: str | None = None
    sponsor_org: str | None = None
    randomization_code_hash: str
    """SHA-256 hex of the randomization code. Never the code itself."""
    created_at: str
    unblinded_by: str | None = None
    unblinded_at: str | None = None
    unblinding_evaluation_id: str | None = None
    emergency_unblinded_by: str | None = None
    emergency_unblinded_at: str | None = None
    suspended_at: str | None = None
    suspended_by: str | None = None

    model_config = {"extra": "allow"}


class ClinicalUnblindingEvent(BaseModel):
    """A single audit event from the clinical unblinding ledger.

    Returned by ``GET /history``. Append-only — a correction is a new event,
    never a mutation of an existing one (§11.10(a) complete records).
    """

    id: str
    org_id: str
    trial_id: str
    blind_id: str
    event_type: ClinicalUnblindingEventType
    actor_id: str
    evaluation_id: str | None = None
    permit_token_hash: str | None = None
    reason: str
    approval_meaning: str | None = None
    """§11.50(a)(2) electronic signature meaning text."""
    subject_ids: list[str] | None = None
    emergency: bool = False
    unblinding_scope: str
    occurred_at: str
    metadata: dict[str, object] | None = None

    model_config = {"extra": "allow"}


class ClinicalBlindRequest(BaseModel):
    """``POST /blind`` — establish a new blinding record for a clinical trial."""

    trial_id: str
    trial_name: str
    phase: str
    blinding_type: str
    randomization_code_hash: str
    """SHA-256 hex hash of the randomization code. Never the code itself."""
    established_by: str
    reason: str
    protocol_number: str | None = None
    evaluation_id: str | None = None
    approval_meaning: str | None = None
    """§11.50(a)(2) electronic signature meaning."""
    sponsor_org: str | None = None

    model_config = {"extra": "allow"}


class ClinicalBlindResponse(BaseModel):
    """``POST /blind`` response."""

    blind: ClinicalTrialBlind

    model_config = {"extra": "allow"}


class ClinicalUnblindRequest(BaseModel):
    """``POST /unblind`` — planned (non-emergency) unblinding of a trial.

    Server requires ``approval_meaning`` ≥ 20 characters (§11.50(a)(2)).
    """

    trial_id: str
    actor_id: str
    reason: str
    approval_meaning: str
    """§11.50(a)(2) electronic signature meaning; server requires ≥ 20 chars."""
    evaluation_id: str | None = None
    permit_token_hash: str | None = None
    unblinding_scope: str | None = None
    """Scope of unblinding (e.g. ``"full"``, ``"partial"``, ``"individual"``)."""
    dsmb_authorization_ref: str | None = None
    """DSMB authorization reference for the audit trail."""

    model_config = {"extra": "allow"}


class ClinicalEmergencyRequest(BaseModel):
    """``POST /emergency`` — ICH E6(R2) §4.8 individual-patient emergency unblinding."""

    trial_id: str
    actor_id: str
    subject_id: str
    emergency_justification: str
    evaluation_id: str | None = None
    permit_token_hash: str | None = None
    approval_meaning: str | None = None
    """§11.50(a)(2) electronic signature meaning."""

    model_config = {"extra": "allow"}


class ClinicalMutationResponse(BaseModel):
    """Response body for ``POST /unblind`` and ``POST /emergency``."""

    success: bool
    trial_id: str
    status: ClinicalBlindingStatus | None = None
    subject_id: str | None = None
    """Present on emergency unblinding — the affected subject identifier."""
    event_type: str | None = None
    unblinded_by: str | None = None
    unblinded_at: str | None = None
    occurred_at: str | None = None
    evaluation_id: str | None = None

    model_config = {"extra": "allow"}


class ClinicalTrialListResponse(BaseModel):
    """``GET`` (list) response."""

    trials: list[ClinicalTrialBlind]

    model_config = {"extra": "allow"}


class ClinicalTrialGetResponse(BaseModel):
    """``GET ?trial_id=X`` response."""

    trial: ClinicalTrialBlind

    model_config = {"extra": "allow"}


class ClinicalHistoryResponse(BaseModel):
    """``GET /history`` response."""

    events: list[ClinicalUnblindingEvent]

    model_config = {"extra": "allow"}


# ---------------------------------------------------------------------------
# Pure helpers
# ---------------------------------------------------------------------------


def is_unblinded(status: ClinicalBlindingStatus) -> bool:
    """Return ``True`` when the trial's randomized assignment has been revealed.

    Both planned (``unblinded``) and emergency (``emergency_unblinded``)
    terminal states count as unblinded. ``blinded``,
    ``unblinding_in_progress`` and ``suspended`` are still blinded.
    """
    return status in _UNBLINDED_STATES


def latest_unblinding_event(
    events: list[ClinicalUnblindingEvent],
) -> ClinicalUnblindingEvent | None:
    """Return the most recent event by ``occurred_at``, or ``None`` if empty.

    Ties resolve to the last element in the supplied order (the ledger is
    already returned newest-or-oldest first by the server; this helper does
    not assume a particular order).
    """
    if not events:
        return None
    return max(events, key=lambda e: e.occurred_at)
